image_writers_guide/guidelines.adoc

```diff
@@ -5,143 +5,186 @@
+Providing environment variables allows consumers of your image to customize behavior, such as database settings, passwords, and performance tuning, without having to introduce a new layer on top of your image. Instead, they can simply define environment variable values when defining a pod and change those settings without rebuilding the image.
 
-=== Configuration
-Users of your image should be able to configure it without having to create a downstream image based on your image.  This means that runtime configuration should be handled via environment variables.  For simple configuration, the running process can consume the environment variables directly.  For more complicated configuration or for runtimes which do not support this, you should configure the runtime by defining a template configuration file which is processed during startup.  During this processing, values supplied via environment variable can be substituted into the configuration file or used to make decisions about what options to set in the configuration file.
+For extremely complex scenarios, configuration can also be supplied using volumes which are expected to be mounted into the container at runtime. If you take this path, your image should provide clear error messages on startup when the necessary volume or configuration is not present.
 
-It is even possible (and recommended) to pass secrets such as certificates and keys into the container via environment variable.  This ensures that the secret values do not end up committed in an image and leaked into a docker registry.
+This topic ties in with the link:#using-services[Using Services for Inter-image Communication] topic in that configuration like datasources should be defined in terms of environment variables which provide the service endpoint information. This allows an application to dynamically consume a datasource service that is defined in the OpenShift environment without modifying the application image.
 
-This allows consumers of your image to customize behavior (e.g. database settings, passwords, performance tuning) without having to introduce a new layer on top of your image.  Instead, they can simply define environment variable values when defining a pod, and change those settings without rebuilding the image.
+In addition, tuning should be done by inspecting the [sysitem]#cgroups# settings for the container. This allows the image to tune itself to the available memory, CPU, and other resources. For example, Java-based images should tune their heap based on the [sysitem]#cgroup# maximum memory parameter to ensure they do not exceed the limits and get an out-of-memory error.
```

Comment:
[Question]
Is "heap' an industry accepted term? I'm not sure, so only asking to confirm and make sure it makes sense to our audience. If we can maybe we can elaborate or clarify this a bit better?